SEED-platform/seed
# 2760 feat/recursive derived column use

#### Any background context you want to provide?
To integrate SEED with BPS there are two features that need to be added to the Derived Column Behavior
1. derived columns can be reordered in the inventory list views 
   - The current functionality places columns first and derived columns second. Derived columns can be reordered within themselves, but will always be listed after the columns. 
3. derived columns can use derived columns in their definitions 
   - The current functionality only allows Columns to be used to define a Derived Column

Both changes require Derived Columns to act like (regular) Columns.

#### What's this PR do?
The primary change is to generate a one to one relationship between the `Column` model and the `DerivedColumn` model. This allows a `DerivedColumn` to be replaced with its related `Column` when being displayed. When it needs to be evaluated the logic will trace the `Column` to its `DerivedColumn` where normal functionality still exists. 

When a `DerivedColumn` is created the logic will automatically generate a `Column` with a property `.derived_column` that will point towards the `DerivedColumn.id`. The existing `Column`s will have a `derived_column` property with a value of `None`.

When defining a derived column its source column may be a `Column` with a related `DerivedColumn`. The logic will traverse the source column tree to determine if the source column contains a reference to the current column. If so an error message will be displayed indicating a circular definition has been created, and the user will be unable to save the derived column.

Now that derived columns may contain a reference to other derived columns, only derived columns that do not contain a reference to a derived column may be deleted. If a user attempts to delete a derived column that contains a reference to another derived column the user will be warned and the derived column will remain.

#### How should this be manually tested?
Any test file will work assuming there is a column with integer values.
You will need to run the new migration

- Define a derived column with another derived column 
    - Create a derived column based on a column with integer values 
    - Create a second derived column that uses the first derived column in its definition (ex: dc2 = dc1 + 10)
    - Confirm values on inventory or beta list

- Test derived columns for circular definitions 
    - to do so, you'll need to create 3 or more derived columns to prove the logics nested capabilities 
    - create a derived column "dc1". It's definition isn't important as we'll be changing it soon.
    - create dc2 = dc1 
    - create dc3 = dc2
    - edit dc1 = dc3 
    - There should be a circular error warning.

- Attempt to delete derived column with definition containing a derived column
    - best to test after testing for circular definitions 
    - attempt to delete dc1 or dc2. Because dc3 is dependent it should prevent a successful deletion. The error message should be clear
    - delete dc3, then dc2, then dc1 successfully.

- Reorder derived columns in the inventory list 
    - generate a few derived columns
    - on the inventory list and beta list click and drag derived columns until they are positioned to the left of regular columns 
    - refresh and confirm they are in the same position.
    - go to column list profile and click and drag derived columns above regular columns 
    - save the profile
    - view the inventory list or beta list and confirm they are in the correct position.



#### What are the relevant tickets?
#2760

#### Screenshots (if appropriate)
